### PR TITLE
env!("OUT_DIR") macro

### PR DESCRIPTION
**Pull request by eupn**
Hello,

This PR is related to the #1964.

@matklad pls give a hint on how do we search for the `OUT_DIR` in the workspace given that we have a `target_path` from `cargo metadata` and can use that.

### CODE REVIEW

File: crates/ra_db/src/input.rs [hunk 1]
```diff
@@ -76,6 +76,7 @@ impl SourceRoot {
 #[derive(Debug, Clone, Default, PartialEq, Eq)]
 pub struct CrateGraph {
     arena: FxHashMap<CrateId, CrateData>,
+    env: FxHashMap<String, String>,
```
Comment (matklad) [inline crates/ra_db/src/input.rs]: That should be per-crate, and not per CrateGraph. That is, it belongs to `CrateData`

Comment (eupn) [reply crates/ra_db/src/input.rs]: Thanks, will propagate that map further down!

File: crates/ra_db/src/input.rs [hunk 2]
```diff
@@ -111,16 +111,28 @@ struct CrateData {
...
+            cfg_options,
+            env: Default::default(),
+        }
     }
 
     fn add_dep(&mut self, name: SmolStr, crate_id: CrateId) {
         self.dependencies.push(Dependency { name, crate_id })
     }
+
+    /// Sets environment variable available trough `env!(VARIABLE_NAME)` macro for crates in crate graph.
```
Comment (kjeremy) [inline crates/ra_db/src/input.rs]: typo: through

**Comment by CAD97** [comment]
> "look at what Cargo does, and do the same"

Highly relevant:

https://github.com/rust-lang/cargo/blob/b03182a8ff0117ffa1c426119ae2b5a418776bbd/src/cargo/core/compiler/context/compilation_files.rs#L216-L220
https://github.com/rust-lang/cargo/blob/b03182a8ff0117ffa1c426119ae2b5a418776bbd/src/cargo/core/compiler/context/compilation_files.rs#L197-L204
https://github.com/rust-lang/cargo/blob/b03182a8ff0117ffa1c426119ae2b5a418776bbd/src/cargo/core/compiler/context/compilation_files.rs#L164-L170

> ```rust
>     /// Returns the "OUT_DIR" directory for running a build script.
>     /// `/path/to/target/{debug,release}/build/PKG-HASH/out`
>     pub fn build_script_out_dir(&self, unit: &Unit<'a>) -> PathBuf {
>         self.build_script_run_dir(unit).join("out")
>     }
> ```

> ```rust
>     /// Returns the directory where a compiled build script is stored.
>     /// `/path/to/target/{debug,release}/build/PKG-HASH`
>     pub fn build_script_dir(&self, unit: &Unit<'a>) -> PathBuf {
>         assert!(unit.target.is_custom_build());
>         assert!(!unit.mode.is_run_custom_build());
>         let dir = self.pkg_dir(unit);
>         self.layout(CompileKind::Host).build().join(dir)
>     }
> ```

> ```rust
>     pub fn pkg_dir(&self, unit: &Unit<'a>) -> String {
>         let name = unit.pkg.package_id().name();
>         match self.metas[unit] {
>             Some(ref meta) => format!("{}-{}", name, meta),
>             None => format!("{}-{}", name, self.target_short_hash(unit)),
>         }
>     }
> ```

`cargo metadata`'s `target_directory` is just the path to the `%workspace/target` directory, so it seems the only way to get `OUT_DIR` would be to guess the package meta disambiguator 😞

If nightly cargo is used, `cargo +nightly build -Z unstable-options --build-plan` outputs a huge JSON blob that includes

```json
{ "invocations": [
  ..,
  { "package_name": "this-package",
    "target_kind": [ "custom-build" ],
    "compile_mode": "run-custom-build",
    "env": { "OUT_DIR": "/absolute/path/to/target/debug/build/this-package-hash/out" }
  },
  ..
}
```

This is present even if the build is fresh and would not be rerun, but does seem to make the build stale and queue a new full rebuild 😞

I think the "best" option would to be adding `OUT_DIR` to the `cargo metadata` information, if the package meta is set at that point. cc https://github.com/rust-lang/cargo/issues/7546

**Comment by CAD97** [comment]
Also related: intellij-rust's implementation of `env("OUT_DIR")`: https://github.com/intellij-rust/intellij-rust/pull/4542

They use `cargo build --build-plan`.

**Comment by edwin0cheng** [comment]
@eupn are you still interested in this patch? I'd be happy to continue to implement it.

**Comment by eupn** [comment]
@edwin0cheng go ahead!

**Comment by CAD97** [comment]
Current status:

OUT_DIR for a given build script is reported in the build script's build-script-executed message. https://github.com/rust-lang/cargo/pull/7622

Here's IntelliJ Rust's (currently off-by-default) support for OUT_DIR: https://github.com/intellij-rust/intellij-rust/pull/4734 (unless they've done more recently, ofc).

**Comment by edwin0cheng** [comment]
@CAD97

Yeah, and the current status of RA part :

We have implemented basic `include` and `concat` eager builtin macro. So this is the final puzzle for supporting `include!(concat!(env!(“OUT_DIR”), “xxx.rs”))!!

**Comment by edwin0cheng** [comment]
Another update :

#3549 is landed ! Although it didn't implement fetching `OUT_DIR` from `cargo check` [yet](https://github.com/rust-analyzer/rust-analyzer/pull/3549#issuecomment-597514225), but I think it works functionally. 

I will close this PR now and feel free to reopen it if needed. 
Thanks @eupn.

**Pull request closed without merging**